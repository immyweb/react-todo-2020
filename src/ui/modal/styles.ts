import glamorous from 'glamorous';

import Button from '../buttons/buttons';
import { colours } from '../../styles/colours';
import { mediaQueries } from '../../styles/breakpoints';
import { fonts } from '../../styles/fonts';

import { rgba, transparentize } from 'polished';

interface IGProps {
  open?: boolean;
  height?: string;
  innerTaller?: boolean;
}

export const ModalContainer = glamorous.div<IGProps>(
  {
    label: 'modal-container',
    position: 'fixed',
    width: '100%',
    height: '100%',
    backgroundColor: rgba(colours.$black, 0.5),
    top: 0,
    left: 0,
    zIndex: 160,
  },
  ({ open }) => ({
    visibility: open ? 'visible' : 'hidden',
  }),
);

export const ModalInner = glamorous.div<IGProps>(
  {
    label: 'modal-inner',
    position: 'relative',
    height: '95%',
    width: '90%',
    top: '50%',
    transform: 'translateY(-50%)',
    background: colours.$white,
    margin: '0 auto',
    padding: '20px 20px 30px 20px',
    overflowY: 'hidden',
    msOverflowStyle: '-ms-autohiding-scrollbar',
    borderRadius: 2,
    [mediaQueries.tabletPortrait]: {
      width: 620,
    },
  },
  ({ innerTaller }) => ({
    height: innerTaller ? '95%' : 'auto',
    [mediaQueries.tabletPortrait]: {
      height: innerTaller ? '80%' : 'auto',
    },
  }),
);

export const ModalContent = glamorous.div<IGProps>(
  {
    label: 'modal-content',
    position: 'relative',
    overflowY: 'auto',
    msOverflowStyle: '-ms-autohiding-scrollbar',
    height: '100%',
    ':after': {
      content: ' ',
      width: '100%',
      position: 'fixed',
      left: 0,
      bottom: 28,
      background: `linear-gradient(to top, ${transparentize(0, colours.$white)} 20%, ${transparentize(
        1,
        colours.$white,
      )})`,
      zIndex: 161,
    },
  },
  ({ innerTaller }) => ({
    paddingBottom: innerTaller ? 60 : 0,
    ':after': {
      height: innerTaller ? 60 : 0,
    },
  }),
);

export const ModalHeadline = glamorous.h2({
  label: 'modal-headline',
  textTransform: 'uppercase',
  textAlign: 'center',
  fontSize: 24,
  fontFamily: fonts.$theSunHeavyCondensed,
  marginBottom: 20,
  lineHeight: 1.1,
  padding: '0 25px',
  [mediaQueries.tabletPortrait]: {
    padding: 0,
  },
});

export const BtnClose = glamorous.div({
  label: 'modal-btn-close',
  textAlign: 'center',
  position: 'absolute',
  top: 10,
  right: 10,
  height: 28,
  width: 28,
  background: colours.$white,
  borderRadius: '50%',
  ':hover': {
    cursor: 'pointer',
  },
});

export const BtnOpen = glamorous(Button)({
  label: 'modal-btn-open',
});
