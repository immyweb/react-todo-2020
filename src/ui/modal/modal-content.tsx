import * as React from 'react';

import { P } from '../typography/typography';
import { ModalHeadline } from '../modal/styles';
import { Image } from '../../styles/globals';
import { IModalContent } from '../../types/modal-content';

export const ModalContent = (modalContent: IModalContent) => {
  const { title, imageTop, copy, imageBottom } = modalContent;
  return (
    <div className="modal-content">
      <ModalHeadline>{title}</ModalHeadline>
      {imageTop && <Image src={imageTop} alt={title} marginBottom={10} />}
      {copy &&
        copy.map((paragraph: string, i: number) => {
          return (
            <P txtalign={'center'} key={i}>
              {paragraph}
            </P>
          );
        })}
      {imageBottom && <Image src={imageBottom} alt={title} marginBottom={10} />}
    </div>
  );
};
