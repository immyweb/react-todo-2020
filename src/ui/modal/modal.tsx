import * as React from 'react';
import Measure from 'react-measure';

import { ModalContainer, ModalInner, ModalContent, BtnClose, BtnOpen } from './styles';
import { setSwipeEnabled } from '../../utils/modal-fix';

import { ReactComponent as CloseIcon } from '../images/icons/close-solid.svg';

interface IProps {
  content: React.ReactNode;
  standalone: boolean;
  openTxt?: string;
  active?: boolean;
  onCloseModal?: () => void;
}

interface IState {
  showModal: boolean;
  outerHeight: number;
  innerHeight: number;
  innerTaller: boolean;
}

export class Modal extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      showModal: props.active ? props.active : false,
      outerHeight: 0,
      innerHeight: 0,
      innerTaller: false,
    };
  }

  UNSAFE_componentWillMount() {
    setSwipeEnabled(true);
  }

  componentDidMount() {
    this.setState({ outerHeight: window.innerHeight - 40 }); // subtrack padding top/bottom
  }

  componentWillUnmount() {
    this.setState({
      showModal: false,
      innerTaller: false,
      outerHeight: 0,
    });

    setSwipeEnabled(false);
  }

  onClose() {
    if (this.props.onCloseModal) {
      this.props.onCloseModal();
    }
  }

  toggle() {
    this.setState({
      showModal: !this.state.showModal,
    });
  }

  render() {
    const { standalone, openTxt, content } = this.props;
    return (
      <Measure
        bounds
        onResize={(contentRect) => {
          const iHeight = contentRect.bounds!.height;
          this.setState({ innerHeight: iHeight + 110 }); // add padding top/bottom
          if (this.state.innerHeight > this.state.outerHeight) {
            this.setState({ innerTaller: true });
          }
        }}
      >
        {({ measureRef }) => (
          <div>
            {standalone && <BtnOpen onClick={() => this.toggle()}>{openTxt ? openTxt : 'Open'}</BtnOpen>}
            <ModalContainer open={this.state.showModal}>
              <ModalInner innerTaller={this.state.innerTaller}>
                <ModalContent innerTaller={this.state.innerTaller}>
                  <div ref={measureRef}>{content}</div>
                </ModalContent>
                <BtnClose onClick={standalone ? () => this.toggle() : () => this.onClose()}>
                  <CloseIcon />
                </BtnClose>
              </ModalInner>
            </ModalContainer>
          </div>
        )}
      </Measure>
    );
  }
}
