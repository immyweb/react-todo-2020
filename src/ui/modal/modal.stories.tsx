import * as React from 'react';

import { storiesOf } from '@storybook/react';

import { Modal } from './modal';
import { H2, P } from '../typography/typography';

const ModalContent = () => {
  return (
    <div className="modal-content">
      <H2>SNOW IN THE DESERT</H2>
      <P>Explain why the person is in this spot as it is obviously words placed distion.</P>
      <img src="http://via.placeholder.com/580x250" alt="" />
      <P>
        Kirk Pasich, a lawyer for the family, said: “Without the results of toxicology tests, we do
        not know what was going on with Chris—or if any to his demise.
      </P>
      <P>
        “The family believes that if Chris took his life, he did not know what he was doing, and
        that drugs or other was doing, that drugs or other substances may have affected his
        actions.”
      </P>
      <P>
        “Without the results of toxicology tests, we do not know what was going on with Chris—or if
        any substances contributed to his demise.
      </P>
      <P>
        Kirk Pasich, a lawyer for the family, said: “Without the results of toxicology tests, we do
        not know what was going on with Chris—or if any to his demise.
      </P>
    </div>
  );
};

storiesOf('Modal', module).add('Modal', () => (
  <Modal content={<ModalContent />} standalone={true} active={false} />
));
