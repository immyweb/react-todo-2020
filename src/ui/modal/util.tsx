import * as React from 'react';
import * as DOM from 'react-dom';
import { podTracker } from '../../utils/pod-tracker-utils';

import { Modal } from '../modal/modal';

import { IModalContent } from '../../types/modal-content';

export const onCloseModal = (domId: string, title: string, id?: string) => {
  podTracker.click({
    id,
    description: `hotspot modal "${title}" closed`,
  });
  DOM.unmountComponentAtNode(document.getElementById(domId) as any);
};

export const onOpenModal = (
  modalData: IModalContent,
  domId: string,
  ModalContentRef: any,
  id?: string,
) => {
  if (modalData) {
    const content = ModalContentRef(modalData);
    podTracker.click({
      id,
      description: `hotspot modal "${modalData.title}" opened`,
    });
    DOM.render(
      <Modal
        content={content}
        standalone={false}
        active={true}
        onCloseModal={() => onCloseModal(domId, modalData.title, id)}
      />,
      document.getElementById(domId),
    );
  }
};
