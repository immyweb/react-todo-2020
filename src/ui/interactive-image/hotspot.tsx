import * as React from 'react';

import Label from '../label/label';
import PulseLabel from '../pulse/pulse-label';
import { onOpenModal } from '../modal/util';

import { Hotspot } from './styles';
import { IHotspot } from './types';

interface IProps {
  hotspot: IHotspot;
  id?: string;
  hotSpotBkgndColour: string;
  hotSpotTxtColour: string;
  pulseColour: string;
  modalRef: React.ReactNode;
}

export const HotspotItem: React.FC<IProps> = ({
  hotspot,
  id,
  hotSpotBkgndColour,
  hotSpotTxtColour,
  pulseColour,
  modalRef,
}) => {
  const { label, modal } = hotspot;
  return (
    <Hotspot
      top={label.position.top}
      left={label.position.left}
      key={label.text}
      className={'interactive-image-hotspot'}
    >
      <Label
        bkgColour={hotSpotBkgndColour}
        txtColour={hotSpotTxtColour}
        arrowSide={label.arrowSide}
        arrowPosition={label.arrowPosition}
        onClick={() => onOpenModal(modal, 'interactiveImageModal', modalRef, id)}
      >
        {label.text}
      </Label>
      <PulseLabel labelSide={label.arrowSide} labelPosition={label.arrowPosition} colour={pulseColour} />
    </Hotspot>
  );
};
