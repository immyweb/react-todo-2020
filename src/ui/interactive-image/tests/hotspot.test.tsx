import * as React from 'react';
import { shallow } from 'enzyme';

import { HotspotItem } from '../hotspot';
import Label, { ArrowPosition } from '../../label/label';
import { onOpenModal } from '../../modal/util';

jest.mock('../../modal/util', () => ({
  onOpenModal: jest.fn(),
}));

const hotspot = {
  label: {
    text: 'ARMS',
    arrowSide: ArrowPosition.BOTTOM,
    arrowPosition: ArrowPosition.LEFT,
    position: {
      top: '30%',
      left: '65%',
    },
  },
  modal: {
    title: 'Long arms are a turn on',
    imageTop:
      'https://www.thesun.co.uk/wp-content/uploads/2018/04/gettyimages-104790256.jpg?strip=all&w=960',
    copy: [
      'Longer arms make you more attractive, according to the University of New South Wales. Long dimensions in the arms can send a subtle signal that youâ€™re a suitable mate.',
    ],
  },
};

const ModalContentRef = () => {
  return <div />;
};

function shallowRender(props?: any) {
  return shallow(
    <HotspotItem
      hotspot={hotspot}
      id={'hotspot'}
      pulseColour={'#000000'}
      hotSpotTxtColour={'#ffffff'}
      hotSpotBkgndColour={'#000000'}
      modalRef={ModalContentRef}
      {...props}
    />,
  );
}

describe('<HotspotItem />', () => {
  it('renders correctly', () => {
    const wrapper = shallowRender();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders hotspot in correct position', () => {
    const wrapper = shallowRender();
    expect(wrapper.props().top).toEqual('30%');
    expect(wrapper.props().left).toEqual('65%');
  });

  it('renders correct hotspot colours', () => {
    const wrapper = shallowRender();
    const label = wrapper.find(Label);
    expect(label.props().bkgColour).toEqual('#000000');
    expect(label.props().txtColour).toEqual('#ffffff');
  });

  it('calls onOpenModal when user clicks on label', () => {
    const wrapper = shallowRender();
    const label = wrapper.find(Label);
    label.simulate('click');
    expect(onOpenModal).toHaveBeenCalledWith(
      hotspot.modal,
      'interactiveImageModal',
      ModalContentRef,
      'hotspot',
    );
  });
});
