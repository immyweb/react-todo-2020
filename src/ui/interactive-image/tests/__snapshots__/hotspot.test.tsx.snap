// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HotspotItem /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    hotSpotBkgndColour="#000000"
    hotSpotTxtColour="#ffffff"
    hotspot={
      Object {
        "label": Object {
          "arrowPosition": "LEFT",
          "arrowSide": "BOTTOM",
          "position": Object {
            "left": "65%",
            "top": "30%",
          },
          "text": "ARMS",
        },
        "modal": Object {
          "copy": Array [
            "Longer arms make you more attractive, according to the University of New South Wales. Long dimensions in the arms can send a subtle signal that youâ€™re a suitable mate.",
          ],
          "imageTop": "https://www.thesun.co.uk/wp-content/uploads/2018/04/gettyimages-104790256.jpg?strip=all&w=960",
          "title": "Long arms are a turn on",
        },
      }
    }
    id="hotspot"
    modalRef={[Function]}
    pulseColour="#000000"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "ARMS",
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Label
          arrowPosition="LEFT"
          arrowSide="BOTTOM"
          bkgColour="#000000"
          onClick={[Function]}
          txtColour="#ffffff"
        >
          ARMS
        </Label>,
        <PulseLabel
          colour="#000000"
          labelPosition="LEFT"
          labelSide="BOTTOM"
        />,
      ],
      "className": "interactive-image-hotspot",
      "left": "65%",
      "top": "30%",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "arrowPosition": "LEFT",
          "arrowSide": "BOTTOM",
          "bkgColour": "#000000",
          "children": "ARMS",
          "onClick": [Function],
          "txtColour": "#ffffff",
        },
        "ref": null,
        "rendered": "ARMS",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "colour": "#000000",
          "labelPosition": "LEFT",
          "labelSide": "BOTTOM",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "ARMS",
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Label
            arrowPosition="LEFT"
            arrowSide="BOTTOM"
            bkgColour="#000000"
            onClick={[Function]}
            txtColour="#ffffff"
          >
            ARMS
          </Label>,
          <PulseLabel
            colour="#000000"
            labelPosition="LEFT"
            labelSide="BOTTOM"
          />,
        ],
        "className": "interactive-image-hotspot",
        "left": "65%",
        "top": "30%",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "arrowPosition": "LEFT",
            "arrowSide": "BOTTOM",
            "bkgColour": "#000000",
            "children": "ARMS",
            "onClick": [Function],
            "txtColour": "#ffffff",
          },
          "ref": null,
          "rendered": "ARMS",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "colour": "#000000",
            "labelPosition": "LEFT",
            "labelSide": "BOTTOM",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
